src/compiler/program.ts

```diff
@@ -490,23 +490,13 @@ namespace ts {
     }
 
     /**
-     * Determined if source file needs to be re-created even if its text hasn't changed
+     * Determine if source file needs to be re-created even if its text hasn't changed
      */
     function shouldProgramCreateNewSourceFiles(program: Program | undefined, newOptions: CompilerOptions) {
-        // If any of these options change, we can't reuse old source file even if version match
-        // The change in options like these could result in change in syntax tree change
+        // If any compiler options change, we can't reuse old source file even if version match
+        // The change in options like these could result in change in syntax tree or `sourceFile.bindDiagnostics`.
         const oldOptions = program && program.getCompilerOptions();
-        return oldOptions && (
-            oldOptions.target !== newOptions.target ||
-            oldOptions.module !== newOptions.module ||
-            oldOptions.moduleResolution !== newOptions.moduleResolution ||
-            oldOptions.noResolve !== newOptions.noResolve ||
-            oldOptions.jsx !== newOptions.jsx ||
-            oldOptions.allowJs !== newOptions.allowJs ||
-            oldOptions.disableSizeLimit !== newOptions.disableSizeLimit ||
-            oldOptions.baseUrl !== newOptions.baseUrl ||
-            !equalOwnProperties(oldOptions.paths, newOptions.paths)
-        );
+        return !!oldOptions && !isJsonEqual(getSourceFileAffectingCompilerOptions(oldOptions), getSourceFileAffectingCompilerOptions(newOptions));

```

Comment:
Btw you have option.type to determine if you want to do arrayIsEqualTo or equalOwnProperties or just direct equals check.

Reply:
Here's how far I got:
```ts
export function areCompilerOptionsEqual(option: CommandLineOption, a: unknown, b: unknown) {
    switch (option.type) {
        case "boolean":
        case "number":
        case "string":
            return a === b;
        case "list":
            return arraysEqual(a as unknown[], b as unknown[], (aa, bb) => areCompilerOptionsEqual((option as CommandLineOptionOfListType).element, aa, bb));
        case "object":
            return equalOwnProperties(a as MapLike<unknown>, b as MapLike<unknown>, isJsonEqual);
        default:
            something with maps;
    }
}
```
Doesn't seem worth it since I need to call `isJsonEqual` in the nested case anyway, and I need some casts which might not be valid if someone put a wrong value in their tsconfig.